File: src/boxing_coach/analysis/rules/school_adherence.py
"""Rule: is the round being boxed in the chosen national school's game?

Only runs when the drill names a `School`. It reads the round-level features
(work rate, mobility, ground coverage, posture, body work, punch selection) and
surfaces the school's expectations the round falls short of — coaching *toward*
the game the fighter says they're training, in that school's language.

This is guidance, not a technique fault, so its observations are MINOR: real
faults (a dropped guard) outrank a school nudge in the correction list.
"""

from __future__ import annotations

from ...domain.analysis import Observation, Severity
from ..context import AnalysisContext
from ..rule import Rule
from ..schools import profile_for, round_feature_values


class SchoolAdherenceRule(Rule):
    id = "school_adherence"

    def applies_to(self, context: AnalysisContext) -> bool:
        return context.drill.school is not None

    def evaluate(self, context: AnalysisContext) -> list[Observation]:
        school = context.drill.school
        if school is None:
            return []

        profile = profile_for(school)
        values = round_feature_values(
            context.round_profile, context.punches, context.drill.stance
        )
        unmet = profile.unmet(values)

        if not unmet:
            return [
                Observation(
                    rule_id=self.id,
                    category=profile.expectations[0].category,
                    severity=Severity.POSITIVE,
                    coaching_text=f"Boxing a clean {profile.label} round — the game's all there.",
                )
            ]

        return [
            Observation(
                rule_id=self.id,
                category=exp.category,
                severity=Severity.MINOR,
                coaching_text=exp.cue,
            )
            for exp in unmet
        ]

File: app/lib/analysis/rules/school_adherence.dart
import '../context.dart';
import '../round_analysis.dart';
import '../rule.dart';
import '../schools.dart';

/// Rule: is the round being boxed in the chosen national school's game? Mirror
/// of `src/boxing_coach/analysis/rules/school_adherence.py`.
///
/// Only runs when the drill names a [School]. It reads the round-level features
/// and surfaces the school's expectations the round falls short of, in that
/// school's language. Guidance, not a technique fault, so the misses are MINOR —
/// a real fault (a dropped guard) outranks a school nudge.
class SchoolAdherenceRule extends Rule {
  @override
  String get id => 'school_adherence';

  @override
  bool appliesTo(AnalysisContext context) => context.drill.school != null;

  @override
  List<Observation> evaluate(AnalysisContext context) {
    final school = context.drill.school;
    if (school == null) return <Observation>[];

    final profile = schoolProfileFor(school);
    final values = roundFeatureValues(
      context.roundProfile,
      context.punches,
      context.drill.stance,
    );
    final unmet = profile.unmet(values);

    if (unmet.isEmpty) {
      return <Observation>[
        Observation(
          ruleId: id,
          category: profile.expectations.first.category,
          severity: Severity.positive,
          coachingText:
              "Boxing a clean ${profile.label} round — the game's all there.",
        ),
      ];
    }

    return <Observation>[
      for (final exp in unmet)
        Observation(
          ruleId: id,
          category: exp.category,
          severity: Severity.minor,
          coachingText: exp.cue,
        ),
    ];
  }
}
